developer_tools/malloc_dbg.c

```diff
@@ -0,0 +1,252 @@
+ */
+void* malloc_dbg(size_t bytes, int line, const char* filename, const char* functionName)
+{
+	void* ptrToReturn = malloc(bytes);
+	int pos = 0;
+	if (!ptrToReturn)
+		return NULL;
+
+	// We must check atexitCalled value to know if we already called the function
+	if (!atexitCalled)
+	{
+		atexit(printLeaks); // Used to call printLeaks when the program exit
+		atexitCalled = 1;
+	}
+
+	pos = inList(filename, line);
+	if (pos == -1)
+	{
+		// Add a new element in the mem_info list
+		memoryInformation = addMemInfo(memoryInformation, ptrToReturn, bytes, line, filename, functionName);
+		if (!memoryInformation)
+		{
+			free(ptrToReturn);
+			return NULL;
+		}
+	}
+	else
+	{
+		editInfo(pos, bytes);
+	}
+	return ptrToReturn;
+}
+
+/**
+ * @brief calloc_dbg function
+ * @details This function calls calloc and allocates the number of bytes passed in the parameters.
+ * If the allocation succeeds then it add the pointer returned by malloc in the mem_info list.
+ * @returns Pointer returned by calloc if it's valid, NULL otherwhise.
+ */
+void* calloc_dbg(size_t elementCount, size_t elementSize, int line, const char* filename, const char* functionName)
+{
+	void* ptrToReturn = calloc(elementCount, elementSize);
+	if (!ptrToReturn)
+		return NULL;
+
+	// We must check atexitCalled value to know if we already called the function
+	if (!atexitCalled)
+	{
+		atexit(printLeaks); // Used to call printLeaks when the program exit
+		atexitCalled = 1;
+	}
+
+	// Add a new element in the mem_info list
+	memoryInformation = addMemInfo(memoryInformation, ptrToReturn, elementCount * elementSize, line, filename, functionName);
+	if (!memoryInformation)
+	{
+		free(ptrToReturn);
+		return NULL;
+	}
+
+	return ptrToReturn;
+}
+
+/**
+ * @brief free_dbg function
+ * @details This function free the memory pointed by the pointer passed in parameter.
+ * To free this pointer, we loop through the mem_info list and check if we find the pointer.
+ * Once it's found, the pointer is freed and the element is deleted from the list.
+ * @returns Nothing.
+ */
+void free_dbg(void* ptrToFree)
+{
+	mem_info* tmp = memoryInformation;
+	mem_info* toFree = NULL;
+	mem_info* previous = NULL;
+
+	// Check if the head contains the pointer to free
+	if (tmp->ptr == ptrToFree)
+	{
+		toFree = tmp;
+		memoryInformation = tmp->next;
+		free(toFree->ptr);
+		free(toFree);
+		if (memoryInformation)
+			memoryInformation->previous = NULL;
+		return;
+	}
+
+	// We can loop through the list without any problems, the head is not the pointer
+	while (tmp)
+	{
+		if (tmp->ptr == ptrToFree) // If we found the pointer that must be freed
+		{
+			toFree = tmp;
+			tmp = tmp->next;
+			previous = toFree->previous;
+
+			if (previous)
+				previous->next = tmp;
+			if (tmp)
+				tmp->previous = previous;
+
+			free(toFree->ptr);
+			if (toFree == memoryInformation)
+				memoryInformation = NULL;
+			free(toFree);
+			return;
+		}
+		tmp = tmp->next;
+	}
+}
+
+/**
+ * @brief printLeaks function

```

Comment:
Describe what the function does but shortly.
`@details` is used for the long description.

Check out other parts of the code too.